server/src/com/cloud/user/DomainManagerImpl.java

```diff
@@ -273,79 +274,97 @@ public boolean deleteDomain(long domainId, Boolean cleanup) {
+                                new CloudRuntimeException("Delete failed on domain " + domain.getName() + " (id: " + domain.getId() +
+                                    "); Please make sure all users and sub domains have been removed from the domain before deleting");
+                            e.addProxyObject(domain.getUuid(), "domainId");
+                            throw e;
+                        }
+                        _messageBus.publish(_name, MESSAGE_REMOVE_DOMAIN_EVENT, PublishScope.LOCAL, domain);
+                    } else {
+                        rollBackState = true;
+                        String msg = null;
+                        if (!accountsForCleanup.isEmpty()) {
+                            msg = accountsForCleanup.size() + " accounts to cleanup";
+                        } else if (!networkIds.isEmpty()) {
+                            msg = networkIds.size() + " non-removed networks";
+                        } else if (hasDedicatedResources) {
+                            msg = "dedicated resources.";
+                        }
+
+                        CloudRuntimeException e = new CloudRuntimeException("Can't delete the domain yet because it has " + msg);
+                        e.addProxyObject(domain.getUuid(), "domainId");
+                        throw e;
                     }
-
-                    CloudRuntimeException e = new CloudRuntimeException("Can't delete the domain yet because it has " + msg);
-                    e.addProxyObject(domain.getUuid(), "domainId");
-                    throw e;
                 }
-            }
 
-            cleanupDomainOfferings(domain.getId());
-            CallContext.current().putContextParameter(Domain.class, domain.getUuid());
-            return true;
-        } catch (Exception ex) {
-            s_logger.error("Exception deleting domain with id " + domain.getId(), ex);
-            if (ex instanceof CloudRuntimeException)
-                throw (CloudRuntimeException)ex;
-            else
-                return false;
-        } finally {
-            //when success is false
-            if (rollBackState) {
-                s_logger.debug("Changing domain id=" + domain.getId() + " state back to " + Domain.State.Active +
-                    " because it can't be removed due to resources referencing to it");
-                domain.setState(Domain.State.Active);
-                _domainDao.update(domain.getId(), domain);
+                cleanupDomainOfferings(domain.getId());
+                CallContext.current().putContextParameter(Domain.class, domain.getUuid());
+                return true;
+            } catch (Exception ex) {
+                s_logger.error("Exception deleting domain with id " + domain.getId(), ex);
+                if (ex instanceof CloudRuntimeException)
+                    throw (CloudRuntimeException)ex;
+                else
+                    return false;
+            } finally {
+                lock.unlock();
```

Comment:
unlock() should be in the outer try-finally block otherwise there may be scenarios where the lock is never released.